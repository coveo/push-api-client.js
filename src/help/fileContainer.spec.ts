const mockedPut = jest.fn();
jest.mock('axios', () => ({
  default: {put: mockedPut},
}));

import {BatchUpdateDocuments, DocumentBuilder} from '..';
import {
  FileContainerResponse,
  uploadContentToFileContainer,
} from './fileContainer';

describe('#uploadContentToFileContainer', () => {
  const fileContainerResponse: FileContainerResponse = {
    uploadUri: 'https://fake.upload.url',
    fileId: 'file_id',
    requiredHeaders: {foo: 'bar'},
  };

  const batch: BatchUpdateDocuments = {
    addOrUpdate: [new DocumentBuilder('http://some.url', 'Some document')],
    delete: [],
  };

  const doMockAxiosPut = () => {
    mockedPut.mockImplementation(() => Promise.resolve());
  };

  beforeEach(() => {
    doMockAxiosPut();
  });

  it('should perform an PUT request with the right params', async () => {
    await uploadContentToFileContainer(fileContainerResponse, batch);

    expect(mockedPut).toHaveBeenCalledWith(
      'https://fake.upload.url/',
      expect.objectContaining({
        addOrUpdate: expect.arrayContaining([
          expect.objectContaining({
            documentId: 'http://some.url',
          }),
        ]),
        delete: expect.arrayContaining([]),
      }),
      {
        headers: {
          foo: 'bar',
        },
        maxBodyLength: 256e6,
      }
    );
  });

  describe('when an MaxBodyLengthExceededError is thrown', () => {
    beforeEach(() => {
      mockedPut.mockImplementationOnce(() =>
        Promise.reject(new FakeMaxBodyLengthExceededError())
      );
    });

    it('should give some info on how to fix it', async () => {
      await expect(
        uploadContentToFileContainer(fileContainerResponse, batch)
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('when anything else than MaxBodyLengthExceededError is thrown', () => {
    const thrownError = new Error();
    beforeEach(() => {
      mockedPut.mockImplementationOnce(() => Promise.reject(thrownError));
    });

    it('should just bubble up the error', async () => {
      await expect(
        uploadContentToFileContainer(fileContainerResponse, batch)
      ).rejects.toBe(thrownError);
    });
  });
});

class FakeMaxBodyLengthExceededError extends Error {
  code = 'ERR_FR_MAX_BODY_LENGTH_EXCEEDED';
  message = 'some initial message';
}
