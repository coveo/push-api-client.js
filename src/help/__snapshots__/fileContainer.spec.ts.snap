// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#uploadContentToFileContainer when an MaxBodyLengthExceededError is thrown should give some info on how to fix it 1`] = `
"some initial message
Try setting and/or increasing \`axios.defaults.maxBodyLength\`."
`;
